keps/sig-api-machinery/2340-Consistent-reads-from-cache/README.md

```diff
@@ -253,6 +253,34 @@ Since falling back to etcd won't work, we should fail the requests and rely on
 rate limiting to prevent cascading failure.  I.e. `Retry-After` HTTP header (for
 well-behaved clients) and [Priority and Fairness](https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20190228-priority-and-fairness.md).
 
+In order to mitigate such problems, let's present how the system currently works
+in different cases. In addition to that, we add column indicating whether a given
+case will change how watchcache implementation will be handling the request.
+
+| ResourceVersion | ResourceVersionMatch | Continuation      | Limit         | etcd implementation                     | watchcache implementation                          | changed  |
+|-----------------|----------------------|-------------------|---------------|-----------------------------------------|----------------------------------------------------|----------|
+| _unset_         | _unset_              | _unset_           | _unset_ / _N_ | Quorum read request                     | Delegated to etcd                                  | Yes      |
+| _unset_         | _unset_              | _token_           | _unset_ / _N_ | Read request from RV encoded in _token_ | Delegated to etcd                                  |          |
+| _unset_         | _Exact_              | _unset_ / _token_ | _unset_ / _N_ | Fails [validation]                      | Fails [validation]                                 |          |
+| _unset_         | _NotOlderThan_       | _unset_           | _unset_ / _N_ | Quorum read request                     | Delegated to etcd                                  | Yes      |
```

Comment:
So this one of the rows in question.  after this kep, this request will be handled by the watch cache.

The question then becomes how does a user request `unset, NotOlderThan, unset, unset` from etcd.  

One proposal is that in order to request `ResourceVersionMatch=NotOlderThan` served from etcd, the user should request `ResourceVersionMatch=Exact` instead.  Thus `unset, NotOlderThan, unset, unset` becomes `unset, Exact, unset, unset`.

Another proposal is to have the user specify another option and indicate `ServeFromEtcd` or `ServeFromWatchCache`.  Thus to request request `ResourceVersionMatch=NotOlderThan` served from etcd, an additional parameter is added `unset, NotOlderThan, unset, unset, ServeFromEtcd`.

I see key decision points as
1. Is secondary authorization for `ServeFromEtcd` necessary? 
2. Is there value in separately selecting `ServeFromEtcd` or `ServeFromWatchCache` for every combination of options?
3. Is it easier to expand the set of allowable tuples for ResourceVersionMatch than adding a new option?  Consider both clients and servers.
4. Does choosing an obfuscated name for `ServeFromEtcd` make the usage of said option less likely?
5. Does adding a new option provide clearer semantics

My current thoughts:
1. I think past-Wojtek argued convincingly for this to be necessary.
2. I think there is value in doing this for all watch-cache back APIs, so List, Get, and Watch. And I think keeping the "do this" aspect of the query consistent and changing only the `ServerFromX` option.
3. I don't think the ResourceVersionMatch is notably easier server-side.  It doesn't achieve any secondary authorization benefit (or if it did, would be more complex).  Client-side I think it's harder to explain.  I'm not even sure that `unset, NotOlderThan, unset, unset`  is the same as `unset, Exact, unset, unset` or if it would mean something different.  Naively I would expect it to mean something different.
4. I don't think dissuading by naming oddly is effective
5. I do think that adding a new option provides clearer semantics due to the ability to universally apply to all requests and exact parity with all other options.  So questions about "does unset mean the same as Exact here?" and "does NotOlderThan mean unset here?".

Reply:
My thoughts on two of those questions:

1. I agree *some* check that limits explicitly bypassing the cache is needed.
2. Since the reason for the explicit "bypass cache" option is to detect / debug / troubleshoot watch cache behavior, keeping all other options identical and changing only the "bypass cache" option is the only way I see to effectively do that. Otherwise, we don't have the ability to debug / troubleshoot watch cache behavior that only manifests for the `ResourceVersionMatch=NotOlderThan` case, for example.